File: internal/models/flag.go
package models

import (
	"encoding/json"
	"time"
)

// Flag is a flag definition, scoped to a project and shared across all of the
// project's environments.
type Flag struct {
	ID          string `json:"id"`
	ProjectID   string `json:"project_id"`
	Key         string `json:"key"`
	Name        string `json:"name"`
	Description string `json:"description"`
	// Salt is a stable per-flag value that salts rollout bucketing so a context
	// isn't stuck in the same percentage slice across different flags.
	Salt string `json:"-"`
	// ClientSideAvailable gates whether client-kind SDK keys (public, e.g.
	// browser) may see/evaluate this flag. Default false = server-side only.
	ClientSideAvailable bool              `json:"client_side_available"`
	Variations          []json.RawMessage `json:"variations"`
	CreatedAt           time.Time         `json:"created_at"`
	UpdatedAt           time.Time         `json:"updated_at"`
}

// Prerequisite gates a flag on another flag: it passes only if the prerequisite
// flag is on and evaluates to Variation.
type Prerequisite struct {
	Key       string `json:"key"`
	Variation int    `json:"variation"`
}

// FlagConfig is a flag's per-environment targeting configuration.
type FlagConfig struct {
	ID            string `json:"id"`
	FlagID        string `json:"flag_id"`
	EnvironmentID string `json:"environment_id"`

	On            bool               `json:"on"`
	OffVariation  int                `json:"off_variation"`
	Prerequisites []Prerequisite     `json:"prerequisites"`
	Targets       []Target           `json:"targets"`
	Rules         []Rule             `json:"rules"`
	Fallthrough   VariationOrRollout `json:"fallthrough"`

	Version   int       `json:"version"`
	CreatedAt time.Time `json:"created_at"`
	UpdatedAt time.Time `json:"updated_at"`
}

// EvalFlag is a flag definition joined with one environment's config: the
// self-contained unit the evaluator consumes, cached per environment.
type EvalFlag struct {
	EnvironmentID       string
	Key                 string
	Salt                string
	ClientSideAvailable bool
	Variations          []json.RawMessage
	On                  bool
	OffVariation        int
	Prerequisites       []Prerequisite
	Targets             []Target
	Rules               []Rule
	Fallthrough         VariationOrRollout
	Version             int
}

File: web/src/types/flag.ts
import type { Clause } from '@/types/segment'

/**
 * A flag *definition*. Per-environment state (on/off, targeting rules) is a
 * separate resource (FlagConfig) and is not part of this shape — the list screen
 * shows the definition; the on/off toggle belongs to a per-environment view.
 *
 * `variations` are opaque JSON values (boolean, string, number, or object), so
 * they are typed as `unknown[]`.
 */
export interface Flag {
  id: string
  project_id: string
  key: string
  name: string
  description: string
  client_side_available: boolean
  variations: unknown[]
  created_at: string
  updated_at: string
}

/** A flag definition plus its on/off state in a specific environment. */
export interface FlagWithState extends Flag {
  on: boolean
}

/** Payload to create/update a flag definition. Variations are opaque JSON values. */
export interface CreateFlagInput {
  key: string
  name: string
  description?: string
  client_side_available?: boolean
  variations: unknown[]
}

/** A variation index or a percentage rollout — how a rule/fallthrough serves a value. */
export interface VariationOrRollout {
  variation?: number
  rollout?: {
    variations: { variation: number; weight: number }[]
    bucketBy?: string
  }
}

/** A targeting rule: contexts matching all clauses get the served variation. */
export interface FlagRule {
  id?: string
  clauses: Clause[]
  variation?: number
  rollout?: VariationOrRollout['rollout']
}

/**
 * A flag's configuration in one environment: the on/off switch, targeting, and
 * the fallthrough. Written via the semantic-instruction PATCH; read via GET.
 */
/** An individual target: serve a variation to specific context keys. */
export interface Target {
  contextKind: string
  values: string[]
  variation: number
}

export interface FlagConfig {
  on: boolean
  off_variation: number
  fallthrough: VariationOrRollout
  targets: Target[]
  rules: FlagRule[]
  version: number
}

/** A semantic instruction applied to a flag's env config (surgical edit). */
export interface FlagInstruction {
  kind: string
  variation?: number
  contextKind?: string
  values?: string[]
  clauses?: Clause[]
  ruleId?: string
}
